Sphinx conversion sql directory

tconsigny opened the pull request:
Converts files in the sql directory into rst format + examples

tconsigny commented on verticapy/sql/create.py:
```diff
@@ -98,6 +105,43 @@ def create_table(
+                     schema = "public",
+                     dtype = {"name": "VARCHAR(60)", "salary": "FLOAT"})
+
+    .. code-block:: python
```
Couldn't get sql magic to display, so just went with the html file approach. If there's a way for sql magic to work, I'll go in and fix this.

mail4umar replied:
Good solution!

oualib commented:
Very long one. I will need time to review it. It is better to cut it into multiple PRs.

mail4umar commented on verticapy/sql/create.py:
```diff
@@ -98,6 +105,43 @@ def create_table(
+        create_table(table_name = "employees",
+                     schema = "public",
+                     dtype = {"name": "VARCHAR(60)", "salary": "FLOAT"})
+
```
@tconsigny  you are creating the table twice.

tconsigny replied:
I thought when genSQL is set to true the SQL is generated but not executed?

mail4umar commented on verticapy/sql/flex.py:
```diff
@@ -100,6 +120,22 @@ def compute_vmap_keys(
+        import verticapy as vp
+        from verticapy.sql import compute_vmap_keys
+
+        vp.read_json("laliga/*.json", table_name = "laliga", materialize = False)
```
the general idea for the codeblock is that the users can directly copy the code and execute. We should try to refrain from  reading files in the user directory like vp.read_json("laliga/*.json", table_name = "laliga", materialize = False). Instead can we do this using load_laliga?

tconsigny replied:
So when I use load_laliga, I'm getting an error that states "public.laliga" is not a flex table

tconsigny replied:
Probably missing something obvious, but I can't get this to work w/o either using sql magic or creating a JSON file and running read_json. If there's an easier way just comment it here, and I'll switch it over.

mail4umar replied:
I also couldn't think of an easier way. Below is what I got. But we want to avoid creating and referencing files. @oualib  any other idea? Can we directly consume the json string to a vDataFrame?

```
import json
data = {
    "column1": {
        "subcolumn1A": "value1A",
        "subcolumn1B": "value1B"
        },
    "column2": {
        "subcolumn2A": "value2A",
        "subcolumn2B": "value2B"
        }
    }   

json_string = json.dumps(data, indent=4)

with open("nested_columns.json", "w") as json_file:
    json_file.write(str(json_string))

vp.create_schema("temp")
vdf = vp.read_json("nested_columns.json", schema = "temp", table_name = "test", flatten_maps = False)
print(compute_vmap_keys(expr = "temp.test", vmap_col = "column1"))
vp.drop("temp.test")
```

mail4umar commented on verticapy/sql/functions/analytic.py:
```diff
@@ -159,6 +311,30 @@ def count(expr: SQLExpression) -> StringSQL:
+        from verticapy import TableSample
+        import verticapy.stats as st
+
+        df = tablesample({"x": [2, -11, None, 12]}).to_vdf()
```
tablesample should be captial: TableSample

mail4umar commented on verticapy/sql/functions/analytic.py:
```diff
@@ -180,6 +356,34 @@ def lag(expr: SQLExpression, offset: int = 1) -> StringSQL:
+        from verticapy import TableSample
+        import verticapy.stats as st
+
+        df = tablesample({"x": [1, 2, 3, 4],
```
TableSample

mail4umar commented on verticapy/sql/functions/analytic.py:
```diff
@@ -202,6 +406,34 @@ def lead(expr: SQLExpression, offset: int = 1) -> StringSQL:
+        from verticapy import TableSample
+        import verticapy.stats as st
+
+        df = tablesample({"x": [1, 2, 3, 4],
```
another TableSample. @tconsigny  you can check all the tablesamples in the file

mail4umar commented on verticapy/sql/functions/analytic.py:
```diff
@@ -221,6 +453,30 @@ def max(expr: SQLExpression) -> StringSQL:
+
+        df = tablesample({"x": [2, -11, 7, 12]}).to_vdf()
+        # apply the max function
+        df.select([st.max(df["x"])])
```
use full code like below:

mail4umar replied:
df.select([str(st.max(df["x"]))])

mail4umar commented on verticapy/sql/functions/analytic.py:
```diff
@@ -240,6 +496,30 @@ def median(expr: SQLExpression) -> StringSQL:
+
+        df = tablesample({"x": [2, -11, 7, 12]}).to_vdf()
+        # apply the median function
+        df.select([st.median(df["x"])])
```
this also seems to be recurring. you can put str() before such examples. 

df.select([str(st.median(df["x"]))])

mail4umar commented on verticapy/sql/functions/analytic.py:
```diff
@@ -259,6 +539,30 @@ def min(expr: SQLExpression) -> StringSQL:
+
+        df = tablesample({"x": [2, -11, 7, 12]}).to_vdf()
+        # apply the min function
+        df.select([st.min(df["x"])])
```
again here

mail4umar reviewed (changes requested):
There are two recurring bugs:

tablesample -> TableSample

And

you should use str() in the code-block when you use it in ipyhon block

oualib commented:
Please update your PR with the latest merge. Arash corrected the pylint issue.

oualib commented on verticapy/sql/flex.py:
```diff
@@ -47,6 +47,40 @@ def compute_flextable_keys(
+                "subcolumn2A": "value2A",
+                "subcolumn2B": "value2B"
+                }
+            }
```
PEP

oualib commented on verticapy/sql/flex.py:
```diff
@@ -47,6 +47,40 @@ def compute_flextable_keys(
+        @suppress
+        vp.drop("temp.test")
+        vdf = vp.read_json("nested_columns.json", schema = "temp", table_name = "test", flatten_maps = False, materialize = False)
+        compute_flextable_keys(flex_name = "temp.test")
```
Do you display the result? I have the impression that you suppress the block which includes the function. Besides, it would be better if you separate the option into two blocks: initialisation and function usage.

tconsigny replied:
output is displayed

oualib commented on verticapy/sql/flex.py:
```diff
@@ -100,6 +134,38 @@ def compute_vmap_keys(
+                "subcolumn2A": "value2A",
+                "subcolumn2B": "value2B"
+                }
+            }
```
Same:
```
data = {
  "column1": {
      "subcolumn1A": "value1A",
      "subcolumn1B": "value1B"
  },
  "column2": {
      "subcolumn2A": "value2A",
      "subcolumn2B": "value2B"
  }
}
```

oualib commented on verticapy/sql/flex.py:
```diff
@@ -100,6 +134,38 @@ def compute_vmap_keys(
+            json_file.write(str(json_string))
+
+        vp.create_schema("temp")
+        vdf = vp.read_json("nested_columns.json", schema = "temp", table_name = "test", flatten_maps = False)
```
same as above

oualib commented on verticapy/sql/flex.py:
```diff
@@ -139,6 +205,38 @@ def isflextable(table_name: str, schema: str) -> bool:
+
+        vp.create_schema("temp")
+        vdf = vp.read_json("nested_columns.json", schema = "temp", table_name = "test", flatten_maps = False, materialize = False)
+        isflextable(table_name = "test", schema = "temp")
```
same as above

oualib commented on verticapy/sql/flex.py:
```diff
@@ -177,6 +275,38 @@ def isvmap(
+        vp.create_schema("temp")
+        vdf = vp.read_json("nested_columns.json", schema = "temp", table_name = "test", flatten_maps = False)
+        isvmap("temp.test", "column1")
+        vp.drop("temp.test")
```
same as above

oualib commented on verticapy/sql/functions/analytic.py:
```diff
@@ -34,6 +34,30 @@ def avg(expr: SQLExpression) -> StringSQL:
+        from verticapy import TableSample
+        import verticapy.stats as st
+
+        df = TableSample({"x": [2, -11, 7, 12]}).to_vdf()
```
it's better to use directly the vDataFrame

```
vDataFrame({"x": [2, -11, 7, 12]})
```

oualib commented on verticapy/sql/functions/analytic.py:
```diff
@@ -34,6 +34,30 @@ def avg(expr: SQLExpression) -> StringSQL:
+    .. code-block:: python
+
+        from verticapy import TableSample
+        import verticapy.stats as st
```
```
import verticapy.sql.functions as vpf

vpf.
```

oualib commented on verticapy/sql/functions/analytic.py:
```diff
@@ -34,6 +34,30 @@ def avg(expr: SQLExpression) -> StringSQL:
+
+        df = TableSample({"x": [2, -11, 7, 12]}).to_vdf()
+        # apply the avg function
+        df.select([str(st.avg(df["x"]))])
```
it's better to bring the comment out and to create a new block.

oualib commented on verticapy/sql/functions/analytic.py:
```diff
@@ -57,6 +81,30 @@ def bool_and(expr: SQLExpression) -> StringSQL:
     -------
     StringSQL
         SQL string.
```
**For everything below:**

Same as above:

1. change verticapy.stats by verticapy.sql.functions as vpf
2. bring the comment of the function usage out the code (create a new block)
3. change TableSample by vDataFrame (so you have to change your import as well)
4. Try to follow PEP
5. Use a new line after the last import

oualib commented on verticapy/sql/functions/conditional.py:
```diff
@@ -32,16 +32,50 @@ def case_when(*args) -> StringSQL:
     args: SQLExpression
```
**For everything below:**

Same as above:

change verticapy.stats by verticapy.sql.functions as vpf
bring the comment of the function usage out the code (create a new block)
change TableSample by vDataFrame (so you have to change your import as well)
Try to follow PEP
Use a new line after the last import

oualib commented on verticapy/sql/functions/date.py:
```diff
@@ -33,6 +33,32 @@ def date(expr: SQLExpression) -> StringSQL:
     -------
```
**For everything below:**

Same as above:

change verticapy.stats by verticapy.sql.functions as vpf
bring the comment of the function usage out the code (create a new block)
change TableSample by vDataFrame (so you have to change your import as well)
Try to follow PEP
Use a new line after the last import

oualib commented on verticapy/sql/functions/math.py:
```diff
@@ -57,6 +57,32 @@ def apply(func: SQLExpression, *args, **kwargs) -> StringSQL:
     -------
```
**For everything below:**

Same as above:

change verticapy.stats by verticapy.sql.functions as vpf
bring the comment of the function usage out the code (create a new block)
change TableSample by vDataFrame (so you have to change your import as well)
Try to follow PEP
Use a new line after the last import

oualib commented on verticapy/sql/functions/null_handling.py:
```diff
@@ -37,6 +37,32 @@ def coalesce(expr: SQLExpression, *args) -> StringSQL:
     -------
```
**For everything below:**

Same as above:

change verticapy.stats by verticapy.sql.functions as vpf
bring the comment of the function usage out the code (create a new block)
change TableSample by vDataFrame (so you have to change your import as well)
Try to follow PEP
Use a new line after the last import

oualib commented on verticapy/sql/functions/random.py:
```diff
@@ -25,13 +25,39 @@ def random() -> StringSQL:
     -------
```
**For everything below:**

Same as above:

change verticapy.stats by verticapy.sql.functions as vpf
bring the comment of the function usage out the code (create a new block)
change TableSample by vDataFrame (so you have to change your import as well)
Try to follow PEP
Use a new line after the last import

oualib commented on verticapy/sql/functions/regexp.py:
```diff
@@ -45,6 +45,32 @@ def regexp_count(
     -------
```
**For everything below:**

Same as above:

change verticapy.stats by verticapy.sql.functions as vpf
bring the comment of the function usage out the code (create a new block)
change TableSample by vDataFrame (so you have to change your import as well)
Try to follow PEP
Use a new line after the last import

oualib commented on verticapy/sql/functions/string.py:
```diff
@@ -36,6 +36,32 @@ def length(expr: SQLExpression) -> StringSQL:
     -------
```
**For everything below:**

Same as above:

change verticapy.stats by verticapy.sql.functions as vpf
bring the comment of the function usage out the code (create a new block)
change TableSample by vDataFrame (so you have to change your import as well)
Try to follow PEP
Use a new line after the last import

oualib commented on verticapy/sql/geo/functions.py:
```diff
@@ -61,6 +61,69 @@ def coordinate_converter(
+    --------
+    .. code-block:: python
+
+        from verticapy.geo import *
```
try to not use import *

oualib commented on verticapy/sql/geo/functions.py:
```diff
@@ -61,6 +61,69 @@ def coordinate_converter(
+        cities = load_cities()
+        cities["lat"] = "ST_X(geometry)"
+        cities["lon"] = "ST_Y(geometry)"
+        cities
```
display(cities) is better

oualib commented on verticapy/sql/geo/functions.py:
```diff
@@ -61,6 +61,69 @@ def coordinate_converter(
+        from verticapy.geo import *
+        from verticapy.datasets import load_cities
+
+        cities = load_cities()
```
add a comment to say you're extracting X and Y coordinate.

oualib commented on verticapy/sql/geo/functions.py:
```diff
@@ -61,6 +61,69 @@ def coordinate_converter(
+
+        # Converting lon, lat to x, y
+        convert_xy = coordinate_converter(cities, "lon", "lat")
+        convert_xy
```
display

oualib commented on verticapy/sql/geo/functions.py:
```diff
@@ -61,6 +61,69 @@ def coordinate_converter(
+
+        # Converting x, y to lon, lat
+        convert_reverse_xy = coordinate_converter(convert_xy, "lon", "lat", reverse=True)
+        convert_reverse_xy
```
display

oualib commented on verticapy/sql/geo/functions.py:
```diff
@@ -61,6 +61,69 @@ def coordinate_converter(
     -------
```
For everything below (all geo files):
 - Try to avoid import *. Import only the needed function or a module (then module.my_fun).
 - Try to create multiple block and bring the comment out when it is possible.
 - Be consistent: Use `display` when you want to display a result.
 - PEP8 for indentation
 - a small sentence to explain what you do.

oualib commented on verticapy/sql/insert.py:
```diff
@@ -66,6 +66,44 @@ def insert_into(
+        from verticapy.sql import insert_into
+
+        # copy left as default (True): one copy
+        insert_into(table_name = "iris",
```
PEP + it is better to create a new block and put the comment out

oualib commented on verticapy/sql/insert.py:
```diff
@@ -66,6 +66,44 @@ def insert_into(
+                            [4.3, 4.7, 9.6, 1.8, "Iris-virginica"]],)
+
+        # copy set to False: multiple inserts
+        insert_into(table_name = "iris",
```
PEP + it is better to create a new block and put the comment out

oualib commented on verticapy/sql/insert.py:
```diff
@@ -66,6 +66,44 @@ def insert_into(
+
+        # genSQL set to True: SQL for inserting data is generated, but not executed
+        # copy set to False: multiple inserts
+        insert_into(table_name = "iris",
```
PEP + it is better to create a new block and put the comment out

oualib commented on verticapy/sql/insert.py:
```diff
@@ -66,6 +66,44 @@ def insert_into(
+                            [4.3, 4.7, 9.6, 1.8, "Iris-virginica"]],
+                    genSQL=True,)
+
+    .. seealso::
```
@mail4umar is this working?

tconsigny commented on verticapy/sql/create.py:
```diff
@@ -98,6 +105,43 @@ def create_table(
+        %load_ext verticapy.sql
+
+        %%sql
+        SELECT * FROM public.employees;
```
As mentioned last week, this and all other examples are taken exactly from the old website. Not sure what the motivation of the initial example author was.

tconsigny commented on verticapy/sql/drop.py:
```diff
@@ -61,6 +76,22 @@ def drop(
+
+    Examples
+    --------
+    .. ipython:: python
```
this is suppressed

tconsigny commented:
I've updated the imports where necessary, changed from st. to vpf., changed all TableSamples to vDataFrames, and addressed some of the PEP formatting.

As for splitting examples into multiple blocks and converting comments into intro sentences, this is what I was talking about last week in regard to a separate project for improving the examples. Right now, the examples in this PR are equal to or better than the old website. Those two other changes would definitely improve the doc, but it probably requires a separate project. I'm out of bandwidth and don't have time to go in and split up sections + add intros for 100+ examples. Happy to create a follow-up project to go in and improve docstrings w/ things like intro sentences and more sections in the next PI.
